File: find-all-duplicates-in-an-array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        answer = []
        n = len(nums)
        left = 0
        while left < n:
            if nums[left] == left + 1:
                left += 1
            else:
                correctPosition = nums[left] - 1
                if nums[correctPosition] == nums[left]:
                    answer.append(nums[left])
                    left += 1
                elif nums[left] < left + 1:
                    nums[nums[left] - 1] =  nums[left]
                    left += 1                
                else:
                    nums[correctPosition], nums[left] = nums[left], nums[correctPosition]
        return answer 

[4,3,2,7,8,2,3,1]
[7,3,2,4,8,2,3,1]
[3,3,2,4,8,2,7,1]
[2,3,3,4,8,2,7,1]
[3,2,3,4,8,2,7,1]
[3,2,3,4,1,2,7,8]
[3,2,3,4,1,2,7,8]
[1,2,3,4,1,2,7,8]

File: 442-find-all-duplicates-in-an-array/find-all-duplicates-in-an-array.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int size = nums.size();
        vector <int> answer;

        for (int i = 0; i < size;) {
            if (nums[i] == i + 1) {
                i++;
            }
            else {
                int correct = nums[i] - 1;
                if (nums[correct] == nums[i]) {
                    answer.push_back(nums[i]);
                    i += 1;
                } else if (nums[i] < i + 1) {
                    nums[correct] = nums[i];
                    i += 1;
                } else {
                    swap(nums[correct], nums[i]);
                }
            }
        }

        return answer;
    }
};